saleor/order/utils.py

```diff
@@ -65,12 +67,51 @@ def decorator(*args, **kwargs):
                 discount = get_voucher_discount_for_order(order)
             except NotApplicable:
                 discount = zero_money(order.currency)
-            order.discount = discount
-        return func(*args, **kwargs)
+        return func(*args, **kwargs, discount=discount)
 
     return decorator
 
 
+def get_voucher_discount_assigned_to_order(order: Order):
+    return order.discounts.filter(type=OrderDiscountType.VOUCHER).first()
+
+
+def recalculate_order_discounts(order: Order):
+    """Recalculate all order discounts assigned to order."""
+
+    order_discounts = order.discounts.filter(type=OrderDiscountType.MANUAL)
+    for order_discount in order_discounts:
+        current_order_discount = copy.deepcopy(order_discount)
+        current_total = order.total.gross.amount
+
+        update_order_discount_for_order(
+            order,
+            order_discount,
+            force_update=True,
+            # The discount is 0 because the total amount was recaluculated based on
+            # lines price
+            already_applied_discount_amount=Money(Decimal(0), currency=order.currency),
+            save_order=False,
+        )
+        discount_value = order_discount.value
+        discount_type = order_discount.value_type
+        amount = order_discount.amount
+
+        if (
+            (
+                discount_type == DiscountValueType.PERCENTAGE
+                or current_total < discount_value
+            )
+            # No need to add new event when new and old amount of discount is the same
+            and amount != current_order_discount.amount
+        ):
```

Comment:
Maybe some of these boolean expressions could be assigned to variables? The `if` condition would be easier to read.